course4/lecture4_en.md

```diff
@@ -0,0 +1,306 @@
+
+Here is another example for better understanding how to use nested patterns. The `zip` function combines two lists into a new list of pairs like a zipper. The length of the resulting list is the minimum of the lengths of the input lists. Given the lists [1, 2, 3] and [a, b, c, d], the zipped list would be (1, a) (2, b) (3, c).
+
+```moonbit
+fn zip(l1: List[Int], l2: List[Char]) -> List[(Int, Char)] {
+  match (l1, l2) {
+    (Cons(hd, tl), Cons(hd2, tl2)) => Cons((hd, hd2), zip(tl, tl2))
+    _ => Nil
+  }
+}
+```
+
+We define our function with pattern matching. Here, we match a pairs by constructing a tuple and then match the nested tuple pattern, effectively matching both lists simultaneously. If either of the input lists is empty, the result is an empty list. When both lists are non-empty, we get a non-empty result. The first item of the result is a tuple of the two values we take from the inputs, followed by the zipped result of the sublists of both lists. Note that the order of pattern matching is top-down. (If a wildcard is placed at the top, the subsequent patterns will never be matched, and the code will never run. The good news is that MoonBit can detect this and provide warnings. These warnings are advisory and won't prevent compilation, so it's crucial to pay attention to the issues panel in your IDE.)
+
+Lastly, pattern matching is not limited to `match`; it can also be used in data binding. In local definitions, we can use pattern matching expressions to bind corresponding substructures to identifiers. It's essential to note that if value comparison occurs at this point and the match fails, the program will encounter a runtime error, leading to program termination.
+
+# Enumerated Types
+
+Now, let's delve into the enumerated types.
+
+Think about this, how should we represent the union of several possibilities? For example, how do we define a type that represents the set of days from Monday to Sunday? How about defining a type for the outcomes of a coin toss – heads or tails? What about a type to represent the results of integer arithmetic operations, such as a successful result, overflow, or division by zero?
+
+The answer is enumerated types. Enumerated types allow us to define data structures that represent different cases. For example, we define a collection of seven days of the week here and the outcomes of a coin toss.
+
+```moonbit
+enum DaysOfWeek {
+  Monday; Tuesday; Wednesday; Thursday; Friday; Saturday; Sunday
+}
+
+enum Coin {
+  Head
+  Tail
+}
+```
+
+The construction of an enumerated type is as follows:
+
+```moonbit
+enum <type_name> { <variant>; }
+```
+
+Here, each possible variant is a constructor. For instance, `let monday = Monday`, where `Monday` defines the day of the week as Monday. Different enumerated types may cause conflicts because they might use the same names for some cases. In such cases, we distinguish them by adding `<type>::` in front of the constructor.
```

Comment:
```suggestion
Here, each possible variant is a constructor. For instance, `let monday = Monday`, where `Monday` defines the day of the week as Monday. Different enumerated types may cause conflicts because they might use the same names for some cases. In such cases, we distinguish them by adding `<type>::` in front of the constructor, such as `DaysOfWeek::Monday`.
```